Fix avoid polygons again

nilsnolde opened the pull request:
Fixes #3182 

The issue was a little subtle: we register `std::vector<PointLL>` as a `bg::polygon` so the edge's `shape` was interpreted as a polygon geometry and then used to intersect with the avoid_polygons, which I guess is sorta undefined behavior (at least for me).

I also added a `to_geojson()` for debugging as you suggested @kevinkreiser, as a trace log, very helpful..

~~Could be done a bit different, by registering another structure with bg, e.g. `std::deque<PointLL>`, to avoid casting every edge to a linestring. With a few experiments I didn't see any significant performance deviation from master though, but let me know if you'd prefer that.~~ that only works if we register std::vector<PointLL> as linestring, of course bg::intersects needs two geometry types. For polygons we wouldn't need to register our own geometry type, we could use bg::model::polygon, but the Tiles::Intersect method somehow won't take that type as an argument.

nilsnolde commented on src/loki/polygon_search.cc:
```diff
@@ -121,9 +158,11 @@ edges_in_rings(const google::protobuf::RepeatedPtrField<valhalla::Options_Ring>&
         auto edge_info = tile->edgeinfo(edge);
+        // vector<PointLL> is reserved for bg as ring, need to cast to linestring
+        line_bg_t edge_line{edge_info.shape().begin(), edge_info.shape().end()};
         bool intersects = false;
         for (const auto& ring_loc : bin.second) {
-          intersects = bg::intersects(rings_bg[ring_loc], edge_info.shape());
+          intersects = bg::intersects(rings_bg[ring_loc], edge_line);
```
that's the actual change

nilsnolde commented:
@PureW maybe this also fixes the clang issues over at https://github.com/valhalla/valhalla/pull/2750#issuecomment-871672316 ?

kevinkreiser commented on src/loki/polygon_search.cc:
```diff
@@ -41,16 +37,10 @@ ring_bg_t PBFToRing(const valhalla::Options::Ring& ring_pbf) {
   for (const auto& coord : ring_pbf.coords()) {
     new_ring.push_back({coord.lng(), coord.lat()});
   }
+  bg::correct(new_ring);
```
whats that do? fix winding or self intersection?

nilsnolde replied:
jep, our rings are expected as clockwise so it fixes counterclockwise rings. also closes open rings if the rings are typed that way (that we do ourselves actually while parsing the request, could remove that tiny bit now). 

I don't remember why, but I had `correct()` in an early version before merging the initial PR and I took it out. couldn't find any commit or so to figure out why I took it out. but since it compiles and tests pass, I'll leave it in (and remove the bit where we close the polygons in the request).

kevinkreiser commented on src/loki/polygon_search.cc:
```diff
@@ -101,7 +91,7 @@ edges_in_rings(const google::protobuf::RepeatedPtrField<valhalla::Options_Ring>&
     rings_bg.push_back(PBFToRing(ring_pbf));
     const ring_bg_t ring_bg = rings_bg.back();
-    rings_length += GetRingLength(ring_bg);
+    rings_length += bg::perimeter(ring_bg);
```
how does this know to use spherical coordinates and to return meters?

kevinkreiser replied:
oh probably because of `geographic`?

nilsnolde replied:
oops, I should've specified the `Haversine` function above, forgot that. But apparently it does some deductions from the underlying point type, no idea what that exactly defaults too though. So rather be explicit and put the Haversine strategy there.

kevinkreiser commented on src/loki/polygon_search.cc:
```diff
@@ -123,7 +150,8 @@ edges_in_rings(const google::protobuf::RepeatedPtrField<valhalla::Options_Ring>&
         for (const auto& ring_loc : bin.second) {
-          intersects = bg::intersects(rings_bg[ring_loc], edge_info.shape());
+          intersects = bg::intersects(rings_bg[ring_loc],
+                                      line_bg_t(edge_info.shape().begin(), edge_info.shape().end()));
```
> avoid_polygons intersected edges as polygons instead of linestrings

so previously it was treating the edge shape as a polygon rather than a line? treating it as a line should be slightly cheaper but i dont understand how that would make the difference for having an intersection and not. maybe it has to do with those edge shapes making self intersecting rings and the self intersecting parts not counting for intersection?

nilsnolde replied:
so the `ei.shape` has `std::vector<PointLL>` type which we register with bg as ring type so that's how it's treated in the bg::intersects call. But of course as an open ring which is the "undefined behavior" part for me.

nilsnolde replied:
When I printed out the avoided edges from a polygon covering only one way I got back like 20 edges before this change. No idea what bg does internally in a situation like that. Now it only prints those 2 DE the polygon really intersects.

PureW commented:
> @PureW maybe this also fixes the clang issues over at [#2750 (comment)](https://github.com/valhalla/valhalla/pull/2750#issuecomment-871672316) ?

Initially getting one error
```
../src/loki/polygon_search.cc:31:19: error: unused variable 'Haversine' [-Werror,-Wunused-const-variable]
static const auto Haversine = [] {
                  ^
1 error generated.
[17/81] Building CXX object src/meili/CMakeFiles/valhalla-meili.dir/map_matcher.cc.o
```

Fixing that, I still get the error in the unittest:

```
[ RUN      ] AvoidTest.TestAvoidShortcutsTruck                                                                                         ../test/gurka/test_avoids.cc:214: Failure                        
Expected equality of these values:                                                                                                     
  avoid_edges.size()                                                                                                                   
    Which is: 0                                                                                                                        
  4                                                                                                                                    [  FAILED  ] AvoidTest.TestAvoidShortcutsTruck (5 ms)
```

nilsnolde commented:
@kevinkreiser thanks for the review, I added haversine perimeter calculation and removed our own ring closing logic when parsing the request.

@PureW the unused-variable warning is gone, forgot to add that function as a strategy.. No idea what to make of that failing test, maybe I'll try clang one of these days

kevinkreiser commented:
this looks good to me. are we saying that compiling with clang fails the tests though?

nilsnolde commented:
> clang fails the tests

seems so @kevinkreiser . though that's not new with this PR, @PureW reported it a couple of weeks back in that initial PR. Must've somehow been introduced with clang 12.0, no idea what's so compiler-specific about that test.. I'll open another issue about it.

PureW commented:
> this looks good to me. are we saying that compiling with clang fails the tests though?

Yea, [I'm seeing this](https://github.com/valhalla/valhalla/pull/2750#issuecomment-871672316) I haven't dug into why this is though.
